### PR TITLE
implementation of temporal rich club

### PR DESCRIPTION
**Pull request by narnolddd**
### What changes were proposed in this pull request?

Temporal rich club coefficient -- measures the stability of edges connecting the highest degree nodes, see [here](https://www.nature.com/articles/s41567-022-01634-8) for an in-depth definition

### Why are the changes needed?

A useful feature for comparing snapshots of temporal networks

### Does this PR introduce any user-facing change? If yes is this documented?

Yes, I have documented the python wrapper of this which I assume will be the most used one.

### How was this patch tested?

Implemented the toy graph example from the paper.

### Are there any further changes required?

* Doing a bit of cloning which might be avoidable. 
* I am calling the `assert_eq_f64` from the pagerank tests, could this be moved to a more centralised `utils` class? Tbh same with the sliding window struct I have made.

### CODE REVIEW

**Comment by narnolddd** [comment]
This is quite slow on persistent graphs, I will try and understand why.

File: raphtory/src/algorithms/motifs/temporal_rich_club_coefficient.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            let new_edges: HashSet<UndirEdge> = g
+                .subgraph(s_k.clone())
+                .edges()
+                .into_iter()
+                .map(|e| undir_edge(e.src().node, e.dst().node))
+                .collect();
+            new_edges
+        })
+        .into_iter()
+        .reduce(|acc_edges, item_edges| acc_edges.intersection(&item_edges).cloned().collect());
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/motifs/temporal_rich_club_coefficient.rs:89]: possibly worth it to use `into_par_iter` here?

**Review by ljeub-pometry — APPROVED** [review]
Neat algorithm to have :)

File: raphtory/src/algorithms/motifs/temporal_rich_club_coefficient.rs (hunk 1 above)
Comment (narnolddd) [reply raphtory/src/algorithms/motifs/temporal_rich_club_coefficient.rs:89]: Thank you! I'll have a look now.

**Pull request merged**